Uses corrected two pass algorithm in theano.tensor.var

kumarkrishna opened the pull request:
Reference #4651 
Adds argument ```corrected``` to maintain backward compatibility.

slefrancois commented:
test this please
we're testing a jenkins bot

ballasn commented on theano/tensor/basic.py:
```diff
@@ -3167,6 +3170,12 @@ def var(input, axis=None, keepdims=False):
     # return the mean sqr
     v = mean((centered_input ** 2), axis, keepdims=keepdims)
+
+    # use 'corrected_two_pass' algorithm
```
you should add a reference to the tech report http://www.cs.yale.edu/publications/techreports/tr222.pdf

ballasn commented on theano/tensor/tests/test_basic.py:
```diff
@@ -6332,6 +6332,8 @@ def test_var():
     assert numpy.allclose(numpy.var(a_val, axis=2), f(a_val))
 
+    f = function([a], var(a, corrected=True))
+    assert numpy.allclose(numpy.var(a_val), f(a_val))
```
I don't think numpy.var uses the corrected two pass algorithm so you probably to implement the corrected two path algorithm as well for the test.

nouiz commented:
@kumarkrishna Can you finish this PR? I think not much is left and it would be great to close this.

thanks.

kumarkrishna commented:
@nouiz Updated the PR with more tests, fixed merge conflicts. I have also updated the implementation to take into consideration the ```ddof``` parameter. I think the patch is ready for review and merge.

nouiz commented on theano/tensor/basic.py:
```diff
@@ -3179,11 +3179,16 @@ def var(input, axis=None, ddof=0, keepdims=False):
     Notes
     -----
-    It uses the two-pass algorithm for more stable results.
+    By default, uses the two-pass algorithm for more stable results.
```
It don't use it by default. Tell that it optionally support that.

kumarkrishna replied:
> By default, uses the two-pass algorithm

This refers to the usual two-pass algorithm used default for calculating the variance, not the corrected version. I will rewrite to make it more clear.

nouiz commented on theano/tensor/var.py:
```diff
@@ -643,9 +643,10 @@ def mean(self, axis=None, dtype=None, keepdims=False, acc_dtype=None):
                                         acc_dtype=acc_dtype)
 
-    def var(self, axis=None, keepdims=False):
+    def var(self, axis=None, keepdims=False, corrected=False):
```
add the ddof parameter at the same time here.

ballasn commented:
pr looks good to me, @nouiz  I think we can merge thanks,

ballasn commented:
Merged, thanks!
